transport/filter.go

```diff
@@ -0,0 +1,115 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package transport
+
+import "golang.org/x/net/context"
+
+// Filter defines transport-level middleware for Outbounds.
+//
+// Filters MAY
+//
+// - change the context
+// - change the request
+// - change the returned response
+// - handle the returned error
+// - call the given outbound zero or more times
+//
+// Filters MUST
+//
+// - always return a non-nil Response or error.
+// - be thread-safe
+//
+// Filters are re-used across requests and MAY be called multiple times on the
+// same request.
+type Filter interface {
+	Call(ctx context.Context, request *Request, out Outbound) (*Response, error)
+}
+
+// NopFilter is a filter that does not do anything special. It simply calls
+// the underlying Outbound.
+var NopFilter Filter = nopFilter{}
+
+// ApplyFilter applies the given Filter to the given Outbound.
+func ApplyFilter(o Outbound, f Filter) Outbound {
+	return filteredOutbound{o: o, f: f}
+}
+
+// FilterFunc adapts a function into a Filter.
+type FilterFunc func(context.Context, *Request, Outbound) (*Response, error)
+
+// Call for FilterFunc.
+func (f FilterFunc) Call(ctx context.Context, request *Request, out Outbound) (*Response, error) {
+	return f(ctx, request, out)
+}
+
+// ChainFilters combines the given filters in-order into a single Filter.
+func ChainFilters(filters ...Filter) Filter {
+	switch len(filters) {
+	case 0:
+		return NopFilter
+	case 1:
+		return filters[0]
+	default:
+		return filterChain(filters)
+	}
+}
+
+type filteredOutbound struct {
+	o Outbound
+	f Filter
+}
+
+func (fo filteredOutbound) Call(ctx context.Context, request *Request) (*Response, error) {
+	return fo.f.Call(ctx, request, fo.o)
+}
+
+type nopFilter struct{}
+
+func (nopFilter) Call(ctx context.Context, request *Request, out Outbound) (*Response, error) {
+	return out.Call(ctx, request)
+}
+
+// filterChain combines a series of filters into a single Filter.
+type filterChain []Filter
+
+func (fc filterChain) Call(ctx context.Context, request *Request, out Outbound) (*Response, error) {
+	return filterChainExec{
+		Chain: []Filter(fc),
+		Final: out,
+	}.Call(ctx, request)
+}
+
+// filterChainExec adapts a series of filters into an Outbound. It is scoped to
+// a single call of an Outbound and is not thread-safe.
+type filterChainExec struct {
+	Chain []Filter
+	Final Outbound
+}
+
+func (cx filterChainExec) Call(ctx context.Context, request *Request) (*Response, error) {
+	if len(cx.Chain) == 0 {
+		return cx.Final.Call(ctx, request)
+	}
+	return cx.Chain[0].Call(ctx, request, filterChainExec{
```

Comment:
optional: since you're using value types, `cx` is already a copy, so you could do:
```go
next := cx.Chain[0]
cx.Chain = cx.Chain[1:]
next.Call(ctx, request, cx)
```

what you have is also fine -- pick whatever you think is clearer